File: tests/cpp-tests/Classes/ExtensionsTest/CocoStudioComponentsTest/ProjectileController.cpp
#include "ProjectileController.h"
#include "SceneController.h"
#include "EnemyController.h"

using namespace cocos2d;
using namespace cocostudio;

ProjectileController::ProjectileController(void)
{
    _name = "ProjectileController";
}

ProjectileController::~ProjectileController(void)
{
}

bool ProjectileController::init()
{
    return true;
}

void ProjectileController::onEnter()
{
    ComController::onEnter();
    auto winSize = Director::getInstance()->getVisibleSize();
    auto origin = Director::getInstance()->getVisibleOrigin();
    _owner->setPosition( Point(origin.x+20, origin.y+winSize.height/2) );
	_owner->setTag(3);
    auto com = _owner->getParent()->getComponent("SceneController");
    static_cast<SceneController*>(com)->getProjectiles().pushBack(_owner);
}

void ProjectileController::onExit()
{

}

void ProjectileController::update(float delta)
{
    auto com = _owner->getParent()->getComponent("SceneController");
    auto _targets = ((SceneController*)com)->getTargets();
    
    auto projectile = dynamic_cast<Sprite*>(_owner);
    auto projectileRect = Rect(
			projectile->getPosition().x - (projectile->getContentSize().width/2),
			projectile->getPosition().y - (projectile->getContentSize().height/2),
			projectile->getContentSize().width,
			projectile->getContentSize().height);

    std::vector<Node*> targetsToDelete;
    targetsToDelete.reserve(20);
    
    for (const auto& target : _targets)
    {
        Rect targetRect = Rect(
            target->getPosition().x - (target->getContentSize().width/2),
            target->getPosition().y - (target->getContentSize().height/2),
            target->getContentSize().width,
            target->getContentSize().height);

        // if (Rect::RectIntersectsRect(projectileRect, targetRect))
        if (projectileRect.intersectsRect(targetRect))
        {
            targetsToDelete.push_back(target);
        }
    }
    
    for (const auto& t : targetsToDelete)
    {
        static_cast<EnemyController*>(t->getComponent("EnemyController"))->die();
    }
    
    if (!targetsToDelete.empty())
    {
        die();
    }
}

ProjectileController* ProjectileController::create(void)
{
    ProjectileController * pRet = new ProjectileController();
    if (pRet && pRet->init())
    {
        pRet->autorelease();
    }
    else
    {
        CC_SAFE_DELETE(pRet);
    }
	return pRet;
}

void freeFunction( Node *ignore )
{
    log("hello");
}

void ProjectileController::move(float flocationX, float flocationY)
{
    auto winSize = Director::getInstance()->getVisibleSize();
    auto origin = Director::getInstance()->getVisibleOrigin();
    // Determinie offset of location to projectile
	float offX = flocationX - _owner->getPosition().x;
	float offY = flocationY - _owner->getPosition().y;

	// Bail out if we are shooting down or backwards
	if (offX <= 0) return;

	// Ok to add now - we've double checked position
	

	// Determine where we wish to shoot the projectile to
	float realX = origin.x + winSize.width + (_owner->getContentSize().width/2);
	float ratio = offY / offX;
	float realY = (realX * ratio) + _owner->getPosition().y;
	Point realDest = Point(realX, realY);

	// Determine the length of how far we're shooting
	float offRealX = realX - _owner->getPosition().x;
	float offRealY = realY - _owner->getPosition().y;
	float length = sqrtf((offRealX * offRealX) + (offRealY*offRealY));
	float velocity = 480/1; // 480pixels/1sec
	float realMoveDuration = length/velocity;

    auto callfunc = CallFuncN::create(
          CC_CALLBACK_1(
                SceneController::spriteMoveFinished,
                static_cast<SceneController*>( getOwner()->getParent()->getComponent("SceneController")
      ) ) );

	// Move projectile to actual endpoint
	_owner->runAction(
          Sequence::create(
               MoveTo::create(realMoveDuration, realDest),
               callfunc,
               NULL)
          );
}

void ProjectileController::die()
{
    auto com = _owner->getParent()->getComponent("SceneController");
    auto& _projectiles = static_cast<SceneController*>(com)->getProjectiles();
    _projectiles.eraseObject(_owner);
    _owner->removeFromParentAndCleanup(true);
}

File: tests/js-tests/src/CocoStudioTest/ComponentsTest/ProjectileController.js
/****************************************************************************
 Copyright (c) 2008-2010 Ricardo Quesada
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2014 Chukong Technologies Inc.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

var ProjectileController = ccs.ComController.extend({
    ctor: function () {
        this._super();
        this._name = "ProjectileController";
    },
    init: function () {
        return true;
    },

    onEnter: function () {
        var winSize = cc.director.getWinSize();
        this.getOwner().x = 30;
	    this.getOwner().y = winSize.height / 2;
        this.getOwner().tag = 3;
        var com = this.getOwner().parent.getComponent("SceneController");
        com.getProjectiles().push(this.getOwner());
    },

    onExit: function () {
    },

    update: function (dt) {
        var com = this.getOwner().parent.getComponent("SceneController");
        var targets = com.getTargets();

        var projectile = this.getOwner();
        var projectileRect = cc.rect(projectile.x - (projectile.width / 2), projectile.y - (projectile.height / 2), projectile.width, projectile.height);

        var targetsToDelete = [];
        var target = null;
        var targetSize = null;
        for (var i = 0; i < targets.length; i++) {
            target = targets[i];
            var targetRect = cc.rect(target.x - (target.width / 2), target.y - (target.height / 2), target.width, target.height);
            if (cc.rectIntersectsRect(projectileRect, targetRect)) {
                targetsToDelete.push(target);
            }
        }

        for (var i = 0; i < targetsToDelete.length; i++) {
            var target = targetsToDelete[i];
            target.getComponent("EnemyController").die();
        }

        var isDied = targetsToDelete.length;

        if (isDied) {
            this.die();
        }
    },

    move: function (x, y) {
        var winSize = cc.director.getWinSize();

        var offX = x - this.getOwner().x;
        var offY = y - this.getOwner().y;

        if (offX <= 0) return;

        // Determine where we wish to shoot the projectile to
        var realX = winSize.width + (this.getOwner().width / 2);
        var ratio = offY / offX;
        var realY = (realX * ratio) + this.getOwner().y;
        var realDest = cc.p(realX, realY);

        // Determine the length of how far we're shooting
        var offRealX = realX - this.getOwner().x;
        var offRealY = realY - this.getOwner().y;
        var length = Math.sqrt((offRealX * offRealX) + (offRealY * offRealY));
        var velocity = 480 / 1; // 480pixels/1sec
        var realMoveDuration = length / velocity;

        // Move projectile to actual endpoint
        this.getOwner().runAction(cc.sequence(
            cc.moveTo(realMoveDuration, realDest),
            cc.callFunc(function () {
                var sceneController = this.getOwner().parent.getComponent("SceneController");
                sceneController.spriteMoveFinished(this.getOwner());
            }, this)));

    },

    die: function () {
        var com = this.getOwner().parent.getComponent("SceneController");
        var projectiles = com.getProjectiles();
        cc.arrayRemoveObject(projectiles, this.getOwner());
        this.getOwner().removeFromParent(true);
    }

});

ProjectileController.create = function () {
    var controller = new ProjectileController();
    controller.init();
    return controller;
};